artifact/v3/header/header.hpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_ARTIFACT_V3_HEADER_PARSER_HPP
+#define MENDER_ARTIFACT_V3_HEADER_PARSER_HPP
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <unordered_map>
+
+#include <nonstd/optional.hpp>
+
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/io.hpp>
+
+#include <artifact/config.hpp>
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+
+using namespace std;
+
+using nonstd::nullopt;
+using nonstd::optional;
+
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+//
+// +---header-info
+//
+
+enum class Payload {
+	RootfsImage,
+	ModuleImage,
+	EmptyPayload,
+};
+
+using PayloadType = pair<Payload, string>;
+
+struct Provides {
+	string artifact_name;
+	optional<string> artifact_group;
+};
+
+struct Depends {
+	vector<string> device_type;
+	optional<vector<string>> artifact_name;
+	optional<vector<string>> artifact_group;
```

Comment:
Shouldn't this be a vector of string pairs?

Reply:
Hmm, maybe, but the name is known, so doing it like this the structure of extraction will be:

`header.info.depends.artifact_group -> vector<string>` 

as opposed to:

`header.info.depends["artifact_group"] -> vector<string>`

Where I chose the former, as it's type-checked, and hence typo safe.

Reply:
Actually, I meant `artifact_name - artifact_group`  should be pairs, but that's not how it is in the JSON (if I recall correctly what I've seen in the code).

Reply:
Ahh, yeah. Ideally they probably should.